code/part-two/processor/actions/add_response.js

```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+const getAddress = require('../utils/addressing');
+const { decode, encode, reject } = require('../utils/helpers');
+
+
+const addResponse = (context, publicKey, { moji, offer }) => {
+  if (!moji || moji.length === 0) {
+    return reject('No moji specified');
+  }
+
+  if (!offer) {
+    return reject('No offer specified');
+  }
+
+  const signer = getAddress.collection(publicKey);
+
+  return context.getState(moji.concat(signer, offer))
+    .then(state => {
+      if (state[signer].length === 0) {
+        return reject('Signer does not have a collection:', publicKey);
+      }
+
+      if (state[offer].length === 0) {
+        return reject('Specified offer does not exist:', offer);
+      }
+
+      for (let offered of moji) {
+        if (state[offered].length === 0) {
+          return reject('Specified moji does not exist:', offered);
+        }
+      }
+
+      moji = moji.sort();
+      const decodedOffer = decode(state[offer]);
+
+      for (let response of decodedOffer.responses) {
+        if (response.moji.every((m, i) => m === moji[i])) {
+          return reject('A response already exists for moji:', moji);
+        }
+      }
+
+      const signerIsOwner = publicKey === decodedOffer.owner;
+      const decodedMoji = moji.map(address => {
+        return Object.assign({ address }, decode(state[address]))
+      });
+
+      if (decodedOffer.owner === decodedMoji[0].owner) {
+        const message = 'Response moji and offer have same owner:';
+        return reject(message, decodedOffer.owner);
+      }
+
+      const approver = signerIsOwner
+        ? decodedMoji[0].owner
+        : decodedOffer.owner;
+
+      for (let offered of decodedMoji) {
```

Comment:
possible to not use the word `offer` here? my brain is falling apart trying to remember that `let offered` is actually each moji that is responded with

maybe `let responseItem`? 🤔 

(lmk if I'm just entirely misinterpreting/missing the point!)

Reply:
Changed to `moji`, and removed the other uses of `offered` as well.